flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/lineage/LineageUtil.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.flink.connector.kafka.lineage;
+
+import org.apache.flink.api.connector.source.Boundedness;
+import org.apache.flink.connector.kafka.lineage.facets.KafkaTopicListFacet;
+import org.apache.flink.connector.kafka.lineage.facets.KafkaTopicPatternFacet;
+import org.apache.flink.streaming.api.lineage.LineageDataset;
+import org.apache.flink.streaming.api.lineage.LineageDatasetFacet;
+import org.apache.flink.streaming.api.lineage.LineageVertex;
+import org.apache.flink.streaming.api.lineage.SourceLineageVertex;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.stream.Collectors;
+
+/** Utility class with useful methods for managing dataset facets. */
+public class LineageUtil {
+
+    private static final String KAFKA_DATASET_PREFIX = "kafka://";
+    private static final String COMMA = ",";
+    private static final String SEMICOLON = ";";
+
+    /**
+     * Loads facet from any object implementing @link{DatasetFacetProvider} interface.
+     *
+     * @param object
+     * @return
+     */
+    public static List<LineageDatasetFacet> facetsFrom(Object object) {
+        return Optional.of(object)
+                .filter(LineageFacetProvider.class::isInstance)
+                .map(LineageFacetProvider.class::cast)
+                .map(LineageFacetProvider::getDatasetFacets)
+                .orElse(Collections.emptyList());
+    }
+
+    /**
+     * Creates dataset from a list of facets. Uses {@link KafkaTopicListFacet} to extract dataset
+     * name from. Dataset per each element of topic list is created
+     *
+     * @param facets
+     * @return
+     */
+    public static List<LineageDataset> datasetsFrom(
```

Comment:
This whole information flow around the facets looks a bit unclean to me.
Both Source/Sink throw a bunch of information into a list of `LineageDatasetFacet`, then this method is applied to take that list apart and construct the actually intended `LineageDataset`. So we first deliberately lose the information of what the facets are about and then we need to use a lot of (hidden) if-else to extract that information again.

WDYT of replacing the `List<LineageDatasetFacet>` instead with a value class that contains all relevant information:
```java
class KafkaFacet { 
  @Nullable
  String topicPattern;
  @Nullable
  List<String> topicList;
  Properties properties;
  @Nullable
  TypeInformation typeInformation;
}
```

Then you can access all the different pieces of information without the isInstance/cast pattern that you use.
You can then in this method still turn all the pieces of information into separate facets.